### PR TITLE
Fallback to uploader if permissions not granted or there is a webcam …

### PR DESCRIPTION
**Pull request by stefaniacardenas**
# Problem
At the moment, if permissions are denied or there is a problem with the webcam stream, the user sees the upload fallback only the first time the component is mounted. If they click on back, then visit again the face step, they will see an empty camera screen.

# Solution
Show uploader screen if permissions are denied at any point of navigation. 
Update: This PR doesn't handle stream error as this will be fixed on `react-webcam`.

## Checklist
_put `n/a` if item is not relevant to PR changes_

- [n/a] Have the CHANGELOG, README, MIGRATION, RELEASE_GUIDELINES docs been updated?
- [n/a ] Have new automated tests been implemented?
- [n/a] Have new manual tests been written down?
- [x] Have tests passed locally?
- [n/a] Have any new strings been translated?

### CODE REVIEW

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-341-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-341-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-341-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-341-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://341-pr-onfido-sdk-ui-onfido.surge.sh

File: src/components/Camera/index.js [hunk 1]
```diff
@@ -112,6 +113,7 @@ export default class Camera extends React.Component<CameraType> {
 
   componentDidMount () {
     this.webcamMounted()
+    if (this.webcam && !this.webcam.stream) return this.props.onWebcamError()
```
Comment (stephencookdev) [inline src/components/Camera/index.js]: Should you not pass something like `{ message: "No stream found" }`?

File: src/components/Capture/capture.js [hunk 2]
```diff
@@ -134,11 +134,20 @@ class Capture extends Component {
     this.onImageFileSelected(file)
   }
 
-  onWebcamError = () => {
-    this.setState({uploadFallback: true})
+  onWebcamError = (error) => {
```
Comment (stephencookdev) [inline src/components/Capture/capture.js]: Should we not log an error? `console.error` if nothing else?

File: src/components/Capture/capture.js [hunk 3]
```diff
@@ -218,16 +228,21 @@ class Capture extends Component {
...
-  const subTitle = useCapture && isDesktop ? i18n.t(`${copyNamespace}.webcam`) : null
+  const title = (!shouldUseWebcam || permissionsDenied) && i18n.t(`${copyNamespace}.upload_title`) ? i18n.t(`${copyNamespace}.upload_title`)  : i18n.t(`${copyNamespace}.title`)
+  const subTitle = shouldUseWebcam && !permissionsDenied && isDesktop ? i18n.t(`${copyNamespace}.webcam`) : null
   const instructions = i18n.t(`${copyNamespace}.instructions`)
   const parentheses = i18n.t('capture_parentheses')
   return (
-    useCapture ?
-      <Camera {...{i18n, method, title, subTitle, ...other}}/> :
-      <Uploader {...{i18n, instructions, parentheses, title, subTitle, ...other}}/>
+    shouldUseWebcam ?
```
Comment (stephencookdev) [inline src/components/Capture/capture.js]: You could change this double ternary into a single ternary, right?

```jsx
const cannotUseWebcam = permissionDenied || error === 'webcam_error'
return shouldUseWebcam && cannotUseWebcam || !shouldUseWebcam ?
  <Uploader {...} /> :
  <Camera {...} />
```

File: src/components/Camera/index.js (hunk 1 above)
Comment (stefaniacardenas) [reply src/components/Camera/index.js]: I could, yes :)

File: src/components/Capture/capture.js (hunk 2 above)
Comment (stefaniacardenas) [reply src/components/Capture/capture.js]: Yes, that would be useful for devs. As part of another story we will also display the error to the user, if they have granted permission but there is an issue with the stream.

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-341-pr-onfido-sdk-ui-onfido.surge.sh

File: src/components/Capture/capture.js (hunk 2 above)
Comment (stefaniacardenas) [reply src/components/Capture/capture.js]: The `react-webcam` dependency already returns an error here.

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-341-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-341-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-341-pr-onfido-sdk-ui-onfido.surge.sh

File: src/core/store/reducers/globals.js [hunk 4]
```diff
@@ -27,6 +29,10 @@ export default function globals(state = initialState, action) {
       return {...state, mobileConnected: action.payload}
     case constants.ACCEPT_TERMS:
       return {...state, termsAccepted: true}
+    case constants.PERMISSIONS_ASKED:
```
Comment (rfreitas) [inline src/core/store/reducers/globals.js]: is this something safe to save on redux? Can't a permission be reasked on some browsers (even on the same session)?

**Comment by rfreitas** [comment]
Due to the multiple sources of truth with this approach (and considering how hard it is to synchronise them). The recommended approach is to fix the cases where reactWebcam is not calling `onWebcamError` and leave the sdk as it was, since it's behaviour had no clear mistakes.

Therefore this PR should not be merged, but instead replaced with another one that points to an upgraded React Webcam

**Pull request closed without merging**